name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/views/MovingAverageConvergenceDivergence.java

```diff
@@ -0,0 +1,121 @@
+    private double macdSubtrahendSmoothingFactor;
+    private double signalLineSmoothingFactor;
+    private Security security;
+    private ChartInterval interval;
+    private ChartLineSeriesAxes macd;
+    private ChartLineSeriesAxes signalLine;
+
+    public MovingAverageConvergenceDivergence(Security security, ChartInterval interval)
+    {
+        this.security = security;
+        this.interval = Objects.requireNonNull(interval);
+        this.macdMinuendSmoothingFactor = 2.0 / (MACD_MINUEND_PERIOD + 1);
+        this.macdSubtrahendSmoothingFactor = 2.0 / (MACD_SUBTRAHEND_PERIOD + 1);
+        this.signalLineSmoothingFactor = 2.0 / (SIGNAL_LINE_PERIOD + 1);
+        this.macd = new ChartLineSeriesAxes();
+        this.signalLine = new ChartLineSeriesAxes();
+
+        calculateMacd();
+    }
+
+    public ChartLineSeriesAxes getMacdLine()
+    {
+        return macd;
+    }
+
+    public ChartLineSeriesAxes getSignalLine()
+    {
+        return signalLine;
+    }
+
+    private void calculateMacd()
+    {
+        if (security == null)
+            return;
+
+        List<SecurityPrice> prices = security.getPricesIncludingLatest();
+        if (prices == null)
+            return;
+
+        int index = Collections.binarySearch(prices, new SecurityPrice(interval.getStart(), 0),
+                        new SecurityPrice.ByDate());
+
+        if (index < 0)
+            index = -index - 1;
+
+        if (index >= prices.size())
+            return;
+
+        List<LocalDate> datesMacd = new ArrayList<>();
+        List<Double> valuesMacd = new ArrayList<>();
+        List<Double> valuesSignal = new ArrayList<>();
+
+        // seed the running EMAs with the first value
+        double minuendEma = prices.get(index).getValue() / Values.Quote.divider();
+        double subtrahendEma = minuendEma;
+        double signalEma = 0;
+
+        for (SecurityPrice price : prices) // NOSONAR
+        {
+            LocalDate date = price.getDate();
+            if (date.isAfter(interval.getEnd()))
+                break;
+
+            double valueDivided = price.getValue() / Values.Quote.divider();
+
+            minuendEma = (valueDivided * macdMinuendSmoothingFactor)
+                            + (minuendEma * (1 - macdMinuendSmoothingFactor));
+            subtrahendEma = (valueDivided * macdSubtrahendSmoothingFactor)
+                            + (subtrahendEma * (1 - macdSubtrahendSmoothingFactor));
+            double macdValue = minuendEma - subtrahendEma;
+            signalEma = (macdValue * signalLineSmoothingFactor) + (signalEma * (1 - signalLineSmoothingFactor));
+
+            if (date.isBefore(interval.getStart()))
+                continue;
+
+            valuesMacd.add(round(macdValue, 2));
+            valuesSignal.add(round(signalEma, 2));
+            datesMacd.add(date);
+        }
+
+        Date[] timelineDates = TimelineChart.toJavaUtilDate(datesMacd.toArray(new LocalDate[0]));
+        macd.setDates(timelineDates);
```

Comment:
```suggestion
        macd.setDates(TimelineChart.toJavaUtilDate(datesMacd.toArray(new LocalDate[0])));
        signalLine.setDates(macd.getDates());
```

Reply:
After changing the calculation for correctness, this code was changed; MACD and MACDSignal have different start dates. In future there will be a xlsx file with the full MACD and MACDSignal calculation.